File: Stack & Queue/leetcode_20.py

# Problem Link ====>>https://leetcode.com/problems/valid-parentheses/description/

def isValid(s):
    closing_brackets = {')': '(', '}': '{', ']': '['}
    stack = []

    for char in s:
        if char in closing_brackets:
            if not stack or stack[-1] != closing_brackets[char]:
                return False
            stack.pop()
        else:
            stack.append(char)

    return not stack


if __name__ == "__main__":
    s = "()"
    isValid = isValid(s)

    if isValid:
        print(f"The string {s} is valid.")
    else:
        print(f"The string {s} is not valid.")

File: Stack & Queue/leetcode_20.cpp

// Problem Link ====>>https://leetcode.com/problems/valid-parentheses/description/

#include <iostream>
#include <unordered_map>
#include <stack>
using namespace std;

class Solution
{
public:
    bool isValid(string s)
    {
        unordered_map<char, char> closing_brackets = {{')', '('}, {'}', '{'}, {']', '['}};
        stack<char> stk;

        for (char c : s)
        {
            if (closing_brackets.count(c))
            {
                if (stk.empty() || stk.top() != closing_brackets[c])
                {
                    return false;
                }
                stk.pop();
            }
            else
            {
                stk.push(c);
            }
        }

        return stk.empty();
    }
};

int main()
{
    string s = "()";
    Solution solution;

    bool isValid = solution.isValid(s);

    if (isValid)
    {
        cout << "The string " << s << " is valid." << endl;
    }
    else
    {
        cout << "The string " << s << " is not valid." << endl;
    }

    return 0;
}